scalableminds/webknossos
# Segment to Bucket Index for Volume Annotations

Creates an index structure for volume annotations. For every segment id, the list of buckets that contain this segment is stored. This is stored separately in the fossildb and updated on volume annotation mutation.

This lays the basis for new routes calculating properties of the segment volumes, e.g. the volume in voxels. A new route already provides that one. More complex metrics (surface area, sphericity, connected components) may be added later

### TODO
 - [x] create data structure (proto vec3int list)
 - [x] add column family, put and get in fossildb
 - [x] on bucket save, get id set
 - [x] on bucket save, load previous bucket (either from fallback layer or volume), get id set
 - [x] get segment index for affected segments, mutate it, write back
 - [x] add route to compute total segment volume
 - [x] duplicate should update this index too
 - [x] downsampling should update this index too
 - [x] revert to version should update this index too (?)
 - [x] how to handle legacy volume annotations with no segment index? hasSegmentIndex → set to true in duplicate (if no fallback layer)
 - [x] perf optimization: skip previous bucket check for “duplicate”
 - [x] perf optimization: process multiple buckets at once (use local buffer + flush)
 - [x] perf optimization: avoid UnsignedInteger conversion of every voxel
    - [x] for uint32
    - [x] for the other types (ShortStream, ByteStream are not available)
 - [x] remove logging, clean up code
 - [x] merging volume tracings should update this index too
 - [x] add some explaining comments (e.g. to buffer)
 - [x] test some more (including merge)

### Follow-up
 - get fallback layer segment index if available (as computed by future versions of voxelytics?) #7132
 - perf optimization: fossildb multiget/multiput (with transaction management?), compare https://github.com/scalableminds/fossildb/issues/39

### Steps to test:
- Create volume annotation without fallback layer
- brush some, with several ids
- go to route `/tracings/volume/<tracingId>/segmentStatistics/volume?segmentId=1&mag=1&token=<token>`
- should get segment volume in vx
- merging annotations should still work (and create merged segment index if all source tracings have segment index)
- saving volume updates should not be unbearably slow (slower than before, though. I had about 100ms for 32 buckets, which is about thrice the old value)

### Issues:
- fixes #7033

------
